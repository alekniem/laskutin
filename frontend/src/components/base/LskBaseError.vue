<template>
    <div class="red--text font-weight-bold">
        <slot />
    </div>
</template>

<script>
export default {
    name: "LskBaseError",
};
</script>
