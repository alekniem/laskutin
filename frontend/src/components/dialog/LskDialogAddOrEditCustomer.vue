<template>
    <v-dialog max-width="700" persistent v-model="dialogVisible">
        <ValidationObserver #default="{ handleSubmit, invalid }" ref="observer">
            <v-form>
                <v-card>
                    <v-card-title :id="cid + 'Title'">
                        {{ title }}
                    </v-card-title>
                    <v-card-text>
                        <v-row dense>
                            <v-col>
                                <LskInputText
                                    :cid="cid + 'NameInput'"
                                    label="Name"
                                    vid="name"
                                    v-model="inputData.name"
                                />
                            </v-col>
                            <v-col>
                                <LskInputText
                                    :cid="cid + 'AddressLineOneInput'"
                                    label="Address line 1"
                                    vid="address_line_one"
                                    v-model="inputData.address_line_one"
                                />
                            </v-col>
                            <v-col>
                                <LskInputText
                                    :cid="cid + 'AddressLineTwoInput'"
                                    label="Address line 2"
                                    vid="address_line_two"
                                    v-model="inputData.address_line_two"
                                />
                            </v-col>
                        </v-row>
                        <template v-if="otherErrors">
                            <LskBaseError
                                v-for="(error, key) in otherErrors"
                                :key="key"
                            >
                                {{ error | trans }}
                            </LskBaseError>
                        </template>
                    </v-card-text>
                    <v-card-actions>
                        <v-spacer />
                        <LskInputButton
                            @click.native="processCancel"
                            :cid="cid + 'CancelButton'"
                            buttonText="Cancel"
                        />
                        <LskInputButton
                            @click.native="handleSubmit(processSave)"
                            :cid="cid + 'SaveButton'"
                            :disabled="invalid || saveButtonDisabled"
                            buttonText="Save"
                        />
                    </v-card-actions>
                </v-card>
            </v-form>
        </ValidationObserver>
    </v-dialog>
</template>

<script>
import LskDialogAddOrEdit from "@/mixins/LskDialogAddOrEdit";

export default {
    name: "LskDialogAddOrEditCustomer",

    mixins: [LskDialogAddOrEdit],

    props: {
        cid: { type: String, required: true },
    },

    data() {
        return {
            fields: ["name", "address_line_one", "address_line_two"],
        };
    },

    computed: {
        apiActionName() {
            return this.id ? "customerUpdate" : "customerStore";
        },

        title() {
            return this.id
                ? this.trans("Edit customer") + " (" + this.id + ")"
                : this.trans("Add new customer");
        },
    },
};
</script>
