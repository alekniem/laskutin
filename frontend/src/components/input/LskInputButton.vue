<template>
    <v-btn :color="color" :disabled="disabled" :id="cid" depressed small>
        {{ buttonText | trans }}
    </v-btn>
</template>

<script>
export default {
    name: "LskInputButton",

    props: {
        buttonText: { type: String, required: true },
        cid: { type: String, required: true },
        color: { type: String, required: false, default: "primary" },
        disabled: { type: Boolean, required: false, default: false },
    },
};
</script>
