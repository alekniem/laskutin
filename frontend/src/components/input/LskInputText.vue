<template>
    <ValidationProvider
        :customMessages="{ required: trans('Required') }"
        :vid="vid"
        ref="provider"
        rules="required"
        tag="div"
        #default="{ errors, valid }"
    >
        <v-text-field
            @change="$emit('input', $event)"
            :error-messages="errors"
            :id="cid"
            :label="trans(label)"
            :success="valid"
            :value="value"
        />
    </ValidationProvider>
</template>

<script>
import { ValidationProvider } from "vee-validate";

export default {
    name: "LskInputText",

    components: {
        ValidationProvider,
    },

    props: {
        cid: { type: String, required: true },
        label: { type: String, required: true },
        value: { type: [String, Number], required: true },
        vid: { type: String, required: true },
    },
};
</script>
