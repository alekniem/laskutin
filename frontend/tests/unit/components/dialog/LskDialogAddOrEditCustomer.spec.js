import flushPromises from "flush-promises";
import LskBaseError from "@/components/base/LskBaseError";
import LskDialogAddOrEditCustomer from "@/components/dialog/LskDialogAddOrEditCustomer";
import LskInputButton from "@/components/input/LskInputButton";
import LskInputText from "@/components/input/LskInputText";
import { createWrapper } from "UT/lsk_test_helper";

const stubs = {
    LskBaseError,
    LskInputButton,
    LskInputText,
};

describe("LskDialogAddOrEditCustomer", () => {
    it("should match snapshot", async () => {
        const wrapper = createWrapper(LskDialogAddOrEditCustomer, {
            propsData: {
                cid: "TEST_CID",
                showDialog: false,
            },
        });
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot("dialog not visible");
    });

    it("should match snapshot", async () => {
        const wrapper = createWrapper(LskDialogAddOrEditCustomer, {
            propsData: {
                cid: "TEST_CID",
                showDialog: true,
            },
            stubs,
        });
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot("dialog visible");
    });

    it("should match snapshot", async () => {
        const wrapper = createWrapper(LskDialogAddOrEditCustomer, {
            propsData: {
                cid: "TEST_CID",
                showDialog: true,
            },
            stubs,
        });
        await flushPromises();
        wrapper.vm._data.inputData.name = "";
        wrapper.vm._data.inputData.address_line_one = "";
        wrapper.vm._data.inputData.address_line_two = "";
        wrapper.vm._data.otherErrors = ["ERROR ONE", "ERROR TWO"];
        await flushPromises();
        wrapper.vm.$refs.observer.validate();
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot("error messages");
    });

    it("should have correct title (no id given)", async () => {
        const wrapper = createWrapper(LskDialogAddOrEditCustomer, {
            propsData: {
                cid: "TEST_CID",
                showDialog: true,
            },
            stubs,
        });
        await flushPromises();
        expect(wrapper.find(".v-card__title").text()).toBe("Add new customer");
    });

    it("should have correct title (id given)", async () => {
        const wrapper = createWrapper(LskDialogAddOrEditCustomer, {
            propsData: {
                cid: "TEST_CID",
                id: 123,
                showDialog: true,
            },
            stubs,
        });
        await flushPromises();
        expect(wrapper.find(".v-card__title").text()).toBe("Edit customer (123)");
    });
});
