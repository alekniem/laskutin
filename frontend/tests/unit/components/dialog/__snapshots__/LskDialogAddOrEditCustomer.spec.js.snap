// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LskDialogAddOrEditCustomer should match snapshot: dialog not visible 1`] = `
<div role="dialog"
     class="v-dialog__container"
>
</div>
`;

exports[`LskDialogAddOrEditCustomer should match snapshot: dialog visible 1`] = `
<div role="dialog"
     class="v-dialog__container"
>
  <div role="document"
       tabindex="0"
       class="v-dialog__content v-dialog__content--active"
       style="z-index: 202;"
  >
    <transition-stub name="dialog-transition"
                     appear="true"
    >
      <div class="v-dialog v-dialog--active v-dialog--persistent"
           style="transform-origin: center center; max-width: 700px;"
      >
        <span>
          <form novalidate="novalidate"
                class="v-form"
          >
            <div class="v-card v-sheet theme--light">
              <div id="TEST_CIDTitle"
                   class="v-card__title"
              >
                Add new customer
              </div>
              <div class="v-card__text">
                <div class="row row--dense">
                  <div class="col">
                    <div>
                      <div class="v-input theme--light v-text-field v-text-field--is-booted">
                        <div class="v-input__control">
                          <div class="v-input__slot">
                            <div class="v-text-field__slot">
                              <label for="TEST_CIDNameInput"
                                     class="v-label theme--light"
                                     style="left: 0px; position: absolute;"
                              >
                                Name
                              </label>
                              <input id="TEST_CIDNameInput"
                                     type="text"
                              >
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light">
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div class="col">
                    <div>
                      <div class="v-input theme--light v-text-field v-text-field--is-booted">
                        <div class="v-input__control">
                          <div class="v-input__slot">
                            <div class="v-text-field__slot">
                              <label for="TEST_CIDAddressLineOneInput"
                                     class="v-label theme--light"
                                     style="left: 0px; position: absolute;"
                              >
                                Address line 1
                              </label>
                              <input id="TEST_CIDAddressLineOneInput"
                                     type="text"
                              >
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light">
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div class="col">
                    <div>
                      <div class="v-input theme--light v-text-field v-text-field--is-booted">
                        <div class="v-input__control">
                          <div class="v-input__slot">
                            <div class="v-text-field__slot">
                              <label for="TEST_CIDAddressLineTwoInput"
                                     class="v-label theme--light"
                                     style="left: 0px; position: absolute;"
                              >
                                Address line 2
                              </label>
                              <input id="TEST_CIDAddressLineTwoInput"
                                     type="text"
                              >
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light">
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div class="v-card__actions">
                <div class="spacer">
                </div>
                <button type="button"
                        class="v-btn v-btn--has-bg theme--light v-size--small primary"
                        id="TEST_CIDCancelButton"
                >
                  <span class="v-btn__content">
                    Cancel
                  </span>
                </button>
                <button type="button"
                        disabled="disabled"
                        class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--small"
                        id="TEST_CIDSaveButton"
                >
                  <span class="v-btn__content">
                    Save
                  </span>
                </button>
              </div>
            </div>
          </form>
        </span>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`LskDialogAddOrEditCustomer should match snapshot: error messages 1`] = `
<div role="dialog"
     class="v-dialog__container"
>
  <div role="document"
       tabindex="0"
       class="v-dialog__content v-dialog__content--active"
       style="z-index: 202;"
  >
    <transition-stub name="dialog-transition"
                     appear="true"
    >
      <div class="v-dialog v-dialog--active v-dialog--persistent"
           style="transform-origin: center center; max-width: 700px;"
      >
        <span>
          <form novalidate="novalidate"
                class="v-form"
          >
            <div class="v-card v-sheet theme--light">
              <div id="TEST_CIDTitle"
                   class="v-card__title"
              >
                Add new customer
              </div>
              <div class="v-card__text">
                <div class="row row--dense">
                  <div class="col">
                    <div>
                      <div class="v-input v-input--has-state theme--light v-text-field v-text-field--is-booted error--text">
                        <div class="v-input__control">
                          <div class="v-input__slot">
                            <div class="v-text-field__slot">
                              <label for="TEST_CIDNameInput"
                                     class="v-label theme--light error--text"
                                     style="left: 0px; position: absolute;"
                              >
                                Name
                              </label>
                              <input id="TEST_CIDNameInput"
                                     type="text"
                              >
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light error--text"
                                 role="alert"
                            >
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                                <div class="v-messages__message">
                                  Required
                                </div>
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div class="col">
                    <div>
                      <div class="v-input v-input--has-state theme--light v-text-field v-text-field--is-booted error--text">
                        <div class="v-input__control">
                          <div class="v-input__slot">
                            <div class="v-text-field__slot">
                              <label for="TEST_CIDAddressLineOneInput"
                                     class="v-label theme--light error--text"
                                     style="left: 0px; position: absolute;"
                              >
                                Address line 1
                              </label>
                              <input id="TEST_CIDAddressLineOneInput"
                                     type="text"
                              >
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light error--text"
                                 role="alert"
                            >
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                                <div class="v-messages__message">
                                  Required
                                </div>
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div class="col">
                    <div>
                      <div class="v-input v-input--has-state theme--light v-text-field v-text-field--is-booted error--text">
                        <div class="v-input__control">
                          <div class="v-input__slot">
                            <div class="v-text-field__slot">
                              <label for="TEST_CIDAddressLineTwoInput"
                                     class="v-label theme--light error--text"
                                     style="left: 0px; position: absolute;"
                              >
                                Address line 2
                              </label>
                              <input id="TEST_CIDAddressLineTwoInput"
                                     type="text"
                              >
                            </div>
                          </div>
                          <div class="v-text-field__details">
                            <div class="v-messages theme--light error--text"
                                 role="alert"
                            >
                              <transition-group-stub tag="div"
                                                     name="message-transition"
                                                     class="v-messages__wrapper"
                              >
                                <div class="v-messages__message">
                                  Required
                                </div>
                              </transition-group-stub>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="red--text font-weight-bold">
                  ERROR ONE
                </div>
                <div class="red--text font-weight-bold">
                  ERROR TWO
                </div>
              </div>
              <div class="v-card__actions">
                <div class="spacer">
                </div>
                <button type="button"
                        class="v-btn v-btn--has-bg theme--light v-size--small primary"
                        id="TEST_CIDCancelButton"
                >
                  <span class="v-btn__content">
                    Cancel
                  </span>
                </button>
                <button type="button"
                        disabled="disabled"
                        class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--small"
                        id="TEST_CIDSaveButton"
                >
                  <span class="v-btn__content">
                    Save
                  </span>
                </button>
              </div>
            </div>
          </form>
        </span>
      </div>
    </transition-stub>
  </div>
</div>
`;
